### ISSUE TITLE
Non-payload characters are URL encoded

### ISSUE THREAD

**Issue by jeroen-80** [sqlmapproject/sqlmap#1204]
I found a POST parameter that seems to be vulnerable. I feed it to sqlmap like this:
/sqlmap.py -r header.txt -p data <more parameters>
header.txt contains:

POST /script.php HTTP/1.1
Host: myapp.com
..
data=(**)

I expect sqlmap to inject URL encoded payloads in between the "(" and ")". This is the case. However, the "(" and ")" are also URL encoded. The results are the same when I use --prefix "(" --suffix ")". Is this a bug or is there a way to leave the characters in the original request alone? Thanks!

**Comment by stamparm** [comment]
And why ( and ) should not be url encoded? Standard says that those should
be too

**Comment by jeroen-80** [comment]
Miroslav Stampar wrote:

> And why ( and ) should not be url encoded? Standard says that those should
> be too

Just to be sure: my question is not about why do or do not encode ( and ) but about why encoding data (in this case ( and ) but can be any character) that is not part of the sqlmap payload (first character after = and last character in http://host/script?data=(sqlmap payload).

Reason why I want to do this (quote sqlmap wiki ;)

"Depending on parameter placement (e.g. GET) its value could be URL encoded by default. In some cases, back-end web servers do not follow RFC standards and require values to be send in their raw non-encoded form. Use --skip-urlencode in those kind of cases."

So I'm interested in:

- Handling sqlmap payload according to chosen setting (default url encoded or not url encoded if --skip-urlencode is used).
- Directly copying and pasting the rest of the uri *without* encoding.

Cheers,

Jeroen

**Comment by stamparm** [comment]
Before we can do the injection, we have to know what's the original content of the POST data. To know the content itself, we have to urldecode it (if it was urlencoded in the first place). Afterwards, we have to follow the rules and url encode the whole value of tested parameter (even the part that is not directly affected by the SQLI, like parentheses in your case) if the content-type is "application/x-www-form-urlencoded".

I am still not sure why do you require explanation for something that is called standard. If content-type is set to "application/x-www-form-urlencoded", then even parentheses have to be urlencoded. What's so strange about that answer. We process the WHOLE parameter value, not just the part that you pointed with *

**Comment by jeroen-80** [comment]
>Before we can do the injection, we have to know what's the original content of the POST data. To  now the content itself, we have to urldecode it (if it was urlencoded in the first place).

Clear.

>Afterwards, we have to follow the rules and url encode the whole value of tested parameter (even the part that is not directly affected by the SQLI, like parentheses in your case) if the content-type is "application/x-www-form-urlencoded".

I didn't knew this about the Content-Type, thanks for explaining!

>I am still not sure why do you require explanation for something that is called standard.

Reason why I want to send non-compliant data is that many developers do NOT understand standards and implement homebrew parsers, possibly introducing security weaknesses. Let's say that backend code copies data enclosed by '(' and ')', performs a url-decode of the data and uses the resulting string in a SQL query. If the data can also contain '(' or ')', the enclosing characters needs to be left alone, the data characters need to be url-encoded since the code otherwise breaks. So the format is "seperator"url-encode(data)"seperator". Trust me, I'm not making this up :)

>If content-type is set to "application/x-www-form-urlencoded", then even parentheses have to be urlencoded. 

> What's so strange about that answer. We process the WHOLE parameter value, not just the part that you pointed with *

So if I'm correct the options I've got are:
- Using "Content-Type:application/x-www-form-urlencoded": seperators and sqlmap payload are url-encoded.
- Using "--skip-urlencode" seperators nor sqlmap payload are url-encoded.
- Using e.g. "Content-Type:text": seperators nor sqlmap payload are url-encoded.
- Values of "--prefix" and "--suffix" are processed based on the rules above.

I'm looking for the option that url-encodes the payload but leaves the seperators in the original uri alone OR an option that url-encodes the payload according to the rules above and leaves the prefix and suffix alone.

I realize that the this is not the most common case but I hope that you understand why it can be useful.

**Comment by stamparm** [comment]
I would say that you should either adapt the code just for that case or try to use --skip-urlencode (maybe it will work in your case). We can't make new options/changes for something that is non-standard and not happening in real world situations

**Comment by jeroen-80** [comment]
>I would say that you should either adapt the code just for that case or try to use --skip-urlencode (maybe it will work in your case). We can't make new options/changes for something that is non-standard

Will do & I understand. 

> and not happening in real world situations

This *is* an actual real world situation at a client... I agree, for sure non-standard but I guess non-standard is the standard in many companies :)

Modifying the code will be the way forward for me. Thanks for your explanations / thoughts!
